fourier/rfft.go

```diff
@@ -0,0 +1,1151 @@
+		ch3.set(0, k, 1, cr2-ci3)
+		ch3.set(0, k, 2, cr2+ci3)
+	}
+
+	if ido == 1 {
+		return
+	}
+
+	idp2 := ido + 1
+	for k := 0; k < l1; k++ {
+		for i := 2; i < ido; i += 2 {
+			ic := idp2 - (i + 1)
+			tr2 := cc3.at(i-1, 2, k) + cc3.at(ic-1, 1, k)
+			cr2 := cc3.at(i-1, 0, k) + taur*tr2
+			ch3.set(i-1, k, 0, cc3.at(i-1, 0, k)+tr2)
+			ti2 := cc3.at(i, 2, k) - cc3.at(ic, 1, k)
+			ci2 := cc3.at(i, 0, k) + taur*ti2
+			ch3.set(i, k, 0, cc3.at(i, 0, k)+ti2)
+			cr3 := taui * (cc3.at(i-1, 2, k) - cc3.at(ic-1, 1, k))
+			ci3 := taui * (cc3.at(i, 2, k) + cc3.at(ic, 1, k))
+			dr2 := cr2 - ci3
+			dr3 := cr2 + ci3
+			di2 := ci2 + cr3
+			di3 := ci2 - cr3
+			ch3.set(i-1, k, 1, wa1[i-2]*dr2-wa1[i-1]*di2)
+			ch3.set(i, k, 1, wa1[i-2]*di2+wa1[i-1]*dr2)
+			ch3.set(i-1, k, 2, wa2[i-2]*dr3-wa2[i-1]*di3)
+			ch3.set(i, k, 2, wa2[i-2]*di3+wa2[i-1]*dr3)
+		}
+	}
+}
+
+func radb4(ido, l1 int, cc, ch, wa1, wa2, wa3 []float64) {
+	cc3 := newThreeArray(ido, 4, l1, cc)
+	ch3 := newThreeArray(ido, l1, 4, ch)
+
+	for k := 0; k < l1; k++ {
+		tr1 := cc3.at(0, 0, k) - cc3.at(ido-1, 3, k)
+		tr2 := cc3.at(0, 0, k) + cc3.at(ido-1, 3, k)
+		tr3 := cc3.at(ido-1, 1, k) + cc3.at(ido-1, 1, k)
+		tr4 := cc3.at(0, 2, k) + cc3.at(0, 2, k)
+		ch3.set(0, k, 0, tr2+tr3)
+		ch3.set(0, k, 1, tr1-tr4)
+		ch3.set(0, k, 2, tr2-tr3)
+		ch3.set(0, k, 3, tr1+tr4)
+	}
+
+	if ido < 2 {
+		return
+	}
+	if ido > 2 {
+		idp2 := ido + 1
+		for k := 0; k < l1; k++ {
+			for i := 2; i < ido; i += 2 {
+				ic := idp2 - (i + 1)
+				ti1 := cc3.at(i, 0, k) + cc3.at(ic, 3, k)
+				ti2 := cc3.at(i, 0, k) - cc3.at(ic, 3, k)
+				ti3 := cc3.at(i, 2, k) - cc3.at(ic, 1, k)
+				tr4 := cc3.at(i, 2, k) + cc3.at(ic, 1, k)
+				tr1 := cc3.at(i-1, 0, k) - cc3.at(ic-1, 3, k)
+				tr2 := cc3.at(i-1, 0, k) + cc3.at(ic-1, 3, k)
+				ti4 := cc3.at(i-1, 2, k) - cc3.at(ic-1, 1, k)
+				tr3 := cc3.at(i-1, 2, k) + cc3.at(ic-1, 1, k)
+				ch3.set(i-1, k, 0, tr2+tr3)
+				cr3 := tr2 - tr3
+				ch3.set(i, k, 0, ti2+ti3)
+				ci3 := ti2 - ti3
+				cr2 := tr1 - tr4
+				cr4 := tr1 + tr4
+				ci2 := ti1 + ti4
+				ci4 := ti1 - ti4
+				ch3.set(i-1, k, 1, wa1[i-2]*cr2-wa1[i-1]*ci2)
+				ch3.set(i, k, 1, wa1[i-2]*ci2+wa1[i-1]*cr2)
+				ch3.set(i-1, k, 2, wa2[i-2]*cr3-wa2[i-1]*ci3)
+				ch3.set(i, k, 2, wa2[i-2]*ci3+wa2[i-1]*cr3)
+				ch3.set(i-1, k, 3, wa3[i-2]*cr4-wa3[i-1]*ci4)
+				ch3.set(i, k, 3, wa3[i-2]*ci4+wa3[i-1]*cr4)
+			}
+		}
+
+		if ido%2 == 1 {
+			return
+		}
+	}
+	for k := 0; k < l1; k++ {
+		ti1 := cc3.at(0, 1, k) + cc3.at(0, 3, k)
+		ti2 := cc3.at(0, 3, k) - cc3.at(0, 1, k)
+		tr1 := cc3.at(ido-1, 0, k) - cc3.at(ido-1, 2, k)
+		tr2 := cc3.at(ido-1, 0, k) + cc3.at(ido-1, 2, k)
+		ch3.set(ido-1, k, 0, tr2+tr2)
+		ch3.set(ido-1, k, 1, math.Sqrt2*(tr1-ti1))
+		ch3.set(ido-1, k, 2, ti2+ti2)
+		ch3.set(ido-1, k, 3, -math.Sqrt2*(tr1+ti1))
+	}
+}
+
+func radb5(ido, l1 int, cc, ch, wa1, wa2, wa3, wa4 []float64) {
+	const (
+		tr11 = 0.309016994374947
+		ti11 = 0.951056516295154
+		tr12 = -0.809016994374947
```

Comment:
I'm confused. The fortran line is `  1-.809016994374947,.587785252292473/`, which implies `tr12` is incorrect, but there's also a random `1` in the line above it. I've been searching for fortran line continuation information, but haven't found anything. You're sure this is right?

Reply:
The Fortran
```
      DATA TR11,TI11,TR12,TI12 /.309016994374947,.951056516295154,
     1-.809016994374947,.587785252292473/
```
is a line continuation (the 1 is in the continuation column), so if the language did not have its archaic line length limits it would be

```
      DATA TR11,TI11,TR12,TI12 /.309016994374947,.951056516295154,-.809016994374947,.587785252292473/
```

which is expressible as pseudo go as

```
const tr11, ri11, tr12, ri12 = 0.309016994374947, 0.951056516295154, -0.809016994374947, 0.587785252292473
```